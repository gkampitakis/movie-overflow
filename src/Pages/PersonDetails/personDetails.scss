.person_details {
  display: flex;
  width: 70%;
  margin: 100px auto;
  padding-bottom: 40px;

  .details {
    height: 100%;
    flex: 3;
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
    border-bottom-right-radius: 8px;
    border-bottom-left-radius: 8px;

    .info {
      span {
        font-weight: bold;
      }
    }

    .person_header {
      background: #333;
      color: white;
      height: 100px;
      justify-content: space-between;
      display: flex;
      flex-direction: column;
      border-top-left-radius: 8px;
      border-top-right-radius: 8px;

      p {
        margin: 5px;
        font-size: 1rem;
      }
    }

    .overview {
      display: flex;
      flex-direction: row;

      img {
        max-width: 200px;
      }

      div {
        p {
          font-size: 0.9rem;
          text-align: left;
        }

        margin: 10px;
        display: flex;
        vertical-align: middle;
        flex-direction: column;
        justify-content: space-around;
      }
    }
  }

  hr {
    margin: 10px 0;
    border: 0.1px solid #e6e6e6;
  }

  .known_for {
    font-size: 1.2rem;
    font-weight: 600;
    margin: 10px;
  }

  .credits {
    height: 25vh;
    overflow-y: auto;
    display: grid;
    grid-template-columns: auto auto auto;

    &:last-child {
      margin-bottom: 20px;
    }

    .item {
      box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
      width: 90%;
      margin: 10px auto;
      border-radius: 8px;
      cursor: pointer;
      display: flex;
      flex-direction: row;
      max-height: 70px;

      div {
        text-align: center;
        margin: auto;
        h3 {
          font-size: 1rem;
          flex: 1;
        }

        p {
          font-size: 0.8rem;
          color: #333;
        }
      }

      img {
        max-width: 40px;
      }
    }
  }
}
