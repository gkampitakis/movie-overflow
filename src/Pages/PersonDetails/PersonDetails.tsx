import React, { useState, useEffect } from 'react';
import { RouteComponentProps, useHistory } from 'react-router-dom';
import { PersonDetails } from '../../types';
import Placeholder from '../../assets/placeholder.jpeg';
import { Loader } from '../../Components';
import { getPerson } from '../../Api';
import './personDetails.scss';

const imgSrc = (path: string, width: '185' | '92') =>
  path ? `https://image.tmdb.org/t/p/w${width}${path}` : Placeholder;

export default function _PersonDetails(
  props: RouteComponentProps<{ id: string }>
) {
  const history = useHistory();
  const [loading, setLoading] = useState(false);
  const { id } = props.match.params;
  const [personDetails, setPersonDetails] = useState<PersonDetails>();

  const handleRequest = () => {
    setLoading(true);

    getPerson(id)
      .then((data) => {
        setPersonDetails(data);
      })
      .catch((error) => {
        if (error.status === 404) {
          return history.push({
            pathname: '/notFound'
          });
        }
        console.error(error);
      })
      .finally(() => setLoading(false));
  };

  const goTo = (id: string, type: 'movie' | 'tv') => {
    history.push({
      pathname: `/${type}/${id}`
    });
  };

  useEffect(() => {
    handleRequest();
  }, [id]);

  return (
    <section className="person_details">
      {personDetails && (
        <>
          <div className="details">
            <header className="person_header">
              <h1>{personDetails.name}</h1>
              <p>{personDetails.known_for_department}</p>
            </header>
            <article className="overview">
              <img
                src={imgSrc(personDetails.profile_path, '185')}
                alt={personDetails.name}
              />
              <div>
                <p>{personDetails.biography}</p>
                <ul className="info">
                  <li>
                    <span>Birthday:</span> {personDetails.birthday}
                  </li>
                  {personDetails.deathday && (
                    <li>
                      <span>Deathday:</span> {personDetails.deathday}
                    </li>
                  )}
                  <li>
                    <span>Place of birth:</span> {personDetails.place_of_birth}
                  </li>
                  <li>
                    <span>Popularity:</span> {personDetails.popularity}
                  </li>
                </ul>
              </div>
            </article>
            {(!!personDetails.combined_credits.cast.length ||
              !!personDetails.combined_credits.crew.length) && (
              <>
                <hr />
                <h2 className="known_for">Known for:</h2>
                <article className="credits">
                  {personDetails.combined_credits.cast.map((elem) => {
                    const { id, media_type, poster_path, character } = elem;
                    const name = elem.original_title || elem.original_name;
                    return (
                      <div
                        className="item"
                        key={id}
                        onClick={() => goTo(id, media_type)}
                      >
                        <img
                          src={imgSrc(poster_path, '92')}
                          alt={name}
                          loading="lazy"
                        />
                        <div>
                          <h3>{name}</h3>
                          {character && <p>as {character}</p>}
                        </div>
                      </div>
                    );
                  })}
                  {personDetails.combined_credits.crew.map((elem) => {
                    const { id, media_type, poster_path, job } = elem;
                    const name = elem.original_title || elem.original_name;
                    return (
                      <div
                        className="item"
                        key={id}
                        onClick={() => goTo(id, media_type)}
                      >
                        <img
                          src={imgSrc(poster_path, '92')}
                          alt={name}
                          loading="lazy"
                        />
                        <div>
                          <h3>{name}</h3>
                          {job && <p>as {job}</p>}
                        </div>
                      </div>
                    );
                  })}
                </article>
              </>
            )}
          </div>
        </>
      )}
      <Loader loading={loading} />
    </section>
  );
}
