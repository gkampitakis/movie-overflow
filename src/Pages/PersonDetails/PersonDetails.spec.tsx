import React from 'react';
import PersonDetails from '.';
import { render, act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FetchMock } from 'jest-fetch-mock';
import {
  mockDelayedResponseOnce,
  mockHistoryPush,
  mockPerson,
  mockResponseOnce
} from '../../setupTests';

const fetchMock = fetch as FetchMock;
const mockProps = {
  match: { params: { id: 'mockId' } }
} as any;

describe('PersonDetails', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    mockHistoryPush.mockClear();
    mockResponseOnce(mockPerson);
  });

  it('Should show loading', async () => {
    fetchMock.resetMocks();
    mockDelayedResponseOnce({});
    await act(async () => {
      const { container } = render(<PersonDetails {...mockProps} />);

      await screen.findByTestId('loader');

      expect(container).toMatchSnapshot();
    });
  });

  it('Should render consistently', async () => {
    await act(async () => {
      const { container } = render(<PersonDetails {...mockProps} />);

      await screen.findByText(/voldemort/i);

      expect(container).toMatchSnapshot();
    });
  });

  it('Should go to correct page', async () => {
    await act(async () => {
      render(<PersonDetails {...mockProps} />);

      await screen.findByText(/mockName/i);
      const cast = screen.getByText(/voldemort/i);
      const crew = screen.getByText(/wizard/i);

      userEvent.click(cast);
      userEvent.click(crew);

      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/movie/1'
      });
      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/tv/3'
      });
    });
  });

  it('Should redirect to page not found', async () => {
    fetchMock.resetMocks();
    mockResponseOnce('{}', 404);
    await act(async () => {
      render(<PersonDetails {...mockProps} />);

      setImmediate(() => {
        expect(mockHistoryPush).toHaveBeenCalledWith({
          pathname: '/notFound'
        });
      });
    });
  });

  it('Should log in case of api error', async () => {
    fetchMock.resetMocks();
    mockResponseOnce({ message: 'mockError' }, 400);
    const logSpy = jest.spyOn(console, 'error');

    await act(async () => {
      render(<PersonDetails {...mockProps} />);

      setImmediate(() => {
        expect(logSpy).toHaveBeenCalled();
      });
    });
  });

  it('Should render placeholder image', async () => {
    fetchMock.resetMocks();
    const response = { ...mockPerson };
    response.profile_path = '';
    mockResponseOnce(response);

    await act(async () => {
      const { container } = render(<PersonDetails {...mockProps} />);

      await screen.findByText(/mockName/i);

      expect(container).toMatchSnapshot();
    });
  });

  it('Should not render credits column if no data', async () => {
    fetchMock.resetMocks();
    const response = { ...mockPerson };
    response.combined_credits.cast = [];
    response.combined_credits.crew = [];
    mockResponseOnce(response);

    await act(async () => {
      const { container } = render(<PersonDetails {...mockProps} />);

      await screen.findByText(/mockName/i);

      expect(container).toMatchSnapshot();
    });
  });
});
