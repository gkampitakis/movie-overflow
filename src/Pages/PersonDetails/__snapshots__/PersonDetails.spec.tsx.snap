// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonDetails Should not render credits column if no data 1`] = `
<div>
  <section
    class="person_details"
  >
    <div
      class="details"
    >
      <header
        class="person_header"
      >
        <h1>
          mockName
        </h1>
        <p>
          Actor
        </p>
      </header>
      <article
        class="overview"
      >
        <img
          alt="mockName"
          src="https://image.tmdb.org/t/p/w185mock_path"
        />
        <div>
          <p>
            mockBiography
          </p>
          <ul
            class="info"
          >
            <li>
              <span>
                Birthday:
              </span>
               
            </li>
            <li>
              <span>
                Deathday:
              </span>
               
              deathDay
            </li>
            <li>
              <span>
                Place of birth:
              </span>
               
              mockPlace
            </li>
            <li>
              <span>
                Popularity:
              </span>
               
              100
            </li>
          </ul>
        </div>
      </article>
    </div>
  </section>
</div>
`;

exports[`PersonDetails Should render consistently 1`] = `
<div>
  <section
    class="person_details"
  >
    <div
      class="details"
    >
      <header
        class="person_header"
      >
        <h1>
          mockName
        </h1>
        <p>
          Actor
        </p>
      </header>
      <article
        class="overview"
      >
        <img
          alt="mockName"
          src="https://image.tmdb.org/t/p/w185mock_path"
        />
        <div>
          <p>
            mockBiography
          </p>
          <ul
            class="info"
          >
            <li>
              <span>
                Birthday:
              </span>
               
            </li>
            <li>
              <span>
                Deathday:
              </span>
               
              deathDay
            </li>
            <li>
              <span>
                Place of birth:
              </span>
               
              mockPlace
            </li>
            <li>
              <span>
                Popularity:
              </span>
               
              100
            </li>
          </ul>
        </div>
      </article>
      <hr />
      <h2
        class="known_for"
      >
        Known for:
      </h2>
      <article
        class="credits"
      >
        <div
          class="item"
        >
          <img
            alt="mockTitle"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath"
          />
          <div>
            <h3>
              mockTitle
            </h3>
            <p>
              as 
              voldemort
            </p>
          </div>
        </div>
        <div
          class="item"
        >
          <img
            alt="mockTitle2"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath2"
          />
          <div>
            <h3>
              mockTitle2
            </h3>
            <p>
              as 
              potter
            </p>
          </div>
        </div>
        <div
          class="item"
        >
          <img
            alt="original_title3"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath3"
          />
          <div>
            <h3>
              original_title3
            </h3>
            <p>
              as 
              wizard
            </p>
          </div>
        </div>
        <div
          class="item"
        >
          <img
            alt="original_name4"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath4"
          />
          <div>
            <h3>
              original_name4
            </h3>
            <p>
              as 
              director
            </p>
          </div>
        </div>
      </article>
    </div>
  </section>
</div>
`;

exports[`PersonDetails Should render placeholder image 1`] = `
<div>
  <section
    class="person_details"
  >
    <div
      class="details"
    >
      <header
        class="person_header"
      >
        <h1>
          mockName
        </h1>
        <p>
          Actor
        </p>
      </header>
      <article
        class="overview"
      >
        <img
          alt="mockName"
          src="placeholder.jpeg"
        />
        <div>
          <p>
            mockBiography
          </p>
          <ul
            class="info"
          >
            <li>
              <span>
                Birthday:
              </span>
               
            </li>
            <li>
              <span>
                Deathday:
              </span>
               
              deathDay
            </li>
            <li>
              <span>
                Place of birth:
              </span>
               
              mockPlace
            </li>
            <li>
              <span>
                Popularity:
              </span>
               
              100
            </li>
          </ul>
        </div>
      </article>
      <hr />
      <h2
        class="known_for"
      >
        Known for:
      </h2>
      <article
        class="credits"
      >
        <div
          class="item"
        >
          <img
            alt="mockTitle"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath"
          />
          <div>
            <h3>
              mockTitle
            </h3>
            <p>
              as 
              voldemort
            </p>
          </div>
        </div>
        <div
          class="item"
        >
          <img
            alt="mockTitle2"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath2"
          />
          <div>
            <h3>
              mockTitle2
            </h3>
            <p>
              as 
              potter
            </p>
          </div>
        </div>
        <div
          class="item"
        >
          <img
            alt="original_title3"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath3"
          />
          <div>
            <h3>
              original_title3
            </h3>
            <p>
              as 
              wizard
            </p>
          </div>
        </div>
        <div
          class="item"
        >
          <img
            alt="original_name4"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w92mockPath4"
          />
          <div>
            <h3>
              original_name4
            </h3>
            <p>
              as 
              director
            </p>
          </div>
        </div>
      </article>
    </div>
  </section>
</div>
`;

exports[`PersonDetails Should show loading 1`] = `
<div>
  <section
    class="person_details"
  >
    <div
      class="loader"
      data-testid="loader"
    />
  </section>
</div>
`;
